### ISSUE TITLE
Abnormally delayed crash with non-vanilla error message

### ISSUE THREAD

**Issue by chocolate-import** [chocolate-doom/chocolate-doom#85]
> The following bug was [originally reported](https://sourceforge.net/p/chocolate-doom/bugs/85/) on Sourceforge by [porschemonty](https://sourceforge.net/u/porschemonty/profile/), 2010-09-14 16:32:12:

All you need to do is try any non-Plutonia2 savegame after pl2.wad has been loaded, this causes the game to freeze for about 12 seconds (you won't be able to do anything during this time, though background music will still play) , then it crashes to desktop with the following error message "Z_Malloc: failed on allocation of 64 bytes"

With vanilla doom2.exe (from the original Plutonia installation, of course) the crash occurs as expected, but without any delay and displaying "Unknown tclass 3 in savegame" instead.

I'm running [r1964](//github.com/fragglet/chocolate-doom/commit/966089e8bdc5d2cae9c1b9e889c9ca1405d6c29e) on WinXP Pro SP3 32bit.

Regards.

**Comment by CapnClever** [comment]
Out of curiosity, I tested this save through several vanilla executables from DOSBox:

* DOOM.EXE (Ultimate Doom 1.9) exits with "P_UnarchiveSpecials:Unknown tclass 184 in savegame"
* DOOM2.EXE (Doom II 1.9) exits with "Unknown tclass 203 in savegame"
* DOOM2.EXE (TNT, Final Doom 1.9) exits with "P_UnarchiveSpecials:Unknown tclass 216 in savegame" and DOSBox spits out some illegal reads
* DOOM2.EXE (Plutonia, Final Doom 1.9) exits with "P_UnarchiveSpecials:Unknown tclass 16 in savegame" and DOSBox spits out some illegal reads (notably fewer than TNT)

Then I tested this in the win32 release of Chocolate Doom 2.2.1:
* `chocolate-doom.exe -iwad doom.wad` exits with "P_UnarchiveSpecials:Unknown tclass 184 in savegame": good so far
* `chocolate-doom.exe -iwad doom2.wad` exits with "Unknown tclass 3 in savegame": almost right, correct WRT reporter's findings but not with my own
* `chocolate-doom.exe -iwad tnt.wad` exits without error: not desirable, same happened with plutonia.wad

Now my guess, based on previous findings when DOSBox starts throwing around read-ops, is that vanilla is leaking somewhere but doesn't care in time before a "truly fatal" error occurs. Chocolate should probably produce SOME message, though I don't know right now what that message ought to be. According to the debugger the crash happens in `P_UnArchiveSpecials()`, as expected, but in this block:
```
switch(tclass)
{
  ...
  case tc_ceiling:
    saveg_read_pad();
    ceiling = Z_Malloc (sizeof(*ceiling), PU_LEVEL, NULL);
    saveg_read_ceiling_t(ceiling);
    ceiling->sector->specialdata = ceiling; // crashes here
...
```
Seems it thinks `tclass` is an appropriate class, maybe by coincidence, but there's no accompanying data so the read via `saveg_read_ceiling_t()` doesn't work out.

**Comment by fabiangreffrath** [comment]
@CapnClever Are your findings reproducible? I mean, does it always quit with the same "Unknown tclass" error messages if you repeat the tests several times?

I guess it does not in the cases in which DOSBOX reports illegal reads. These should be the cases in which Choco reports "unexpected end of file" and `fread()` stops writing the (uninitialized) `result` variable - with the noticable difference that the DOS version apparently really reads junk beyond the end of the file buffer. Thus, the results should differ in repeated experiments.

**Comment by CapnClever** [comment]
All of the DOSBox-vanilla errors are reproducible. I wouldn't be surprised if it ended up inconsistent on a DOS machine, though, as DOSBox itself may just have a reproducible state. I did try to potentially change up the memory by playing through the games and booting up a few others in the same DOSBox session but they always gave back the same error.

**Comment by fabiangreffrath** [comment]
Thank you for testing! Interesting, so it seems that the value reported by the error message is independent of the stack. This may have to do with the fact that Vanilla Doom indeed reads past the end of the file from the filesystem and doesn't do the intermediate save in a stack variable that Choco does.

Could you please perform another test for me? Please create a copy of the savegame, delete the original one and restore the game from the copy. Please don't just *move* the file, because this will in fact keep it in place and just change its meta-data, but I want to be sure that the savegame is stored at another place in the file system. If this indeed leads to a different error message, my theory is right and Vanilla reads past the file end, which we could impossibly emulate in Choco. We would then have to initialize the `result` variable in `saveg_read8()` with something "random", because apparently stack variable aren't that "random" anymore nowadays (i.e. on modern systems with all their stack protection and probably initialization mechanisms).

**Comment by haleyjd** [comment]
I don't really understand what's being asserted with "reads past the end of the file." fread and other such functions cannot read "past the end" of a disk file, that is not a logically valid concept. In order for standard C file IO to work at all, this is never allowable, and everything I know about Watcom's libc suggests it's also simply untrue.

**Comment by fabiangreffrath** [comment]
Hm, I am not sure either (how about `getc(*f++)` or similar nasty stuff?). Maybe it's really just of matter of the stack being initialized or not.

**Comment by CapnClever** [comment]
Vanilla and Chocolate savegame handling is quite different: vanilla reads the entire file into the buffer and then adjusts the buffer's pointer to access info, while Chocolate reads across the file as it needs more info, accessing at most a few bytes at a time. (I actually "fixed" this for Chocolate Hexen recently, and looking at the original Doom code it was the same back then.) If Chocolate reaches the end of the file, then according to @haleyjd it's definitely not going to attempt to read past it. However, with vanilla's usage, it's entirely possible to read past the buffer.

Anyway, I tried your test @fabiangreffrath:

1. Tried running the savegame (error `P_UnarchiveSpecials:Unknown tclass 255 in savegame`)
2. Used DOSBox for the copy/deletion (namely `copy doomsav4.dsg doomsav5.dsg` and `del doomsav4.dsg`)
3. Tried running the copied savegame (error `P_UnarchiveSpecials:Unknown tclass 255 in savegame`)
